src/typedb_jupyter/query.py

```diff
@@ -0,0 +1,205 @@
+    elif keyword_counts["group"] > 0:
+        candidate_query_types.append("match-group")
+    elif keyword_counts["get"] > 0:
+        candidate_query_types.append("match")
+
+    if keyword_counts["define"] > 0:
+        candidate_query_types.append("define")
+
+    if keyword_counts["undefine"] > 0:
+        candidate_query_types.append("undefine")
+
+    if keyword_counts["insert"] > 0 and keyword_counts["delete"] > 0:
+        candidate_query_types.append("update")
+    elif keyword_counts["insert"] > 0:
+        candidate_query_types.append("insert")
+    elif keyword_counts["delete"] > 0:
+        candidate_query_types.append("delete")
+
+    if len(candidate_query_types) > 1:
+        raise QueryParsingError("Query contains incompatible keywords: '{}'".format("', '".join(candidate_query_types)))
+    elif len(candidate_query_types) == 1:
+        query_type = candidate_query_types[0]
+    elif keyword_counts["match"] > 0:
+        query_type = "match"
+    else:
+        raise QueryParsingError("Query contains no keywords.")
+
+    if args.session is None:
+        if strict_transactions:
+            raise ArgumentError("Strict transaction types is enabled and no session type was provided. Use -s to specify session type.")
+        elif query_type in ("define", "undefine"):
+            session_type = SessionType.SCHEMA
+        else:
+            session_type = SessionType.DATA
+    else:
+        if args.session.lower() == "schema":
+            session_type = SessionType.SCHEMA
+        elif args.session.lower() == "data":
+            session_type = SessionType.DATA
+        else:
+            raise ArgumentError("Incorrect session type provided. Session type must be 'schema' or 'data'.")
+
+    if args.transaction is None:
+        if strict_transactions:
+            raise ArgumentError("Strict transaction types is enabled and no transaction type was provided. Use -t to specify transaction type.")
+        elif query_type in ("define", "undefine", "insert", "update", "delete"):
+            transaction_type = TransactionType.WRITE
+        else:
+            transaction_type = TransactionType.READ
+    else:
+        if args.transaction.lower() == "read":
+            transaction_type = TransactionType.READ
+        elif args.transaction.lower() == "write":
+            transaction_type = TransactionType.WRITE
+        else:
+            raise ArgumentError("Incorrect transaction type provided. Transaction type must be 'read' or 'write'.")
+
+    if args.inference is None:
+        options = typedb.client.TypeDBOptions().core().set_infer(global_inference)
+    else:
+        options = typedb.client.TypeDBOptions().core().set_infer(args.inference)
+
+    connection.set_session(session_type)
+
+    if show_info:
+        print_query_info(connection, session_type, transaction_type, query_type)
+
+    try:
+        with connection.session.transaction(transaction_type, options) as transaction:
```

Comment:
This can't be done. The transaction has to be instantiated only at the time the query is run, as otherwise the session might otherwise no longer exist.